pkg/platform/api/buildlogstream/streamer.go

```diff
@@ -0,0 +1,156 @@
+package buildlogstream
+
+import (
+	"net/http"
+
+	"github.com/go-openapi/strfmt"
+	"github.com/gorilla/websocket"
+
+	"github.com/ActiveState/cli/internal/errs"
+	"github.com/ActiveState/cli/internal/locale"
+	"github.com/ActiveState/cli/internal/logging"
+	"github.com/ActiveState/cli/pkg/platform/api"
+	"github.com/ActiveState/cli/pkg/platform/model"
+)
+
+type Request struct {
+	recipeID   strfmt.UUID
+	msgHandler MessageHandler
+}
+
+func NewRequest(recipeID strfmt.UUID, msgHandler MessageHandler) *Request {
+	return &Request{recipeID: recipeID, msgHandler: msgHandler}
+}
+
+type MessageHandler interface {
+	BuildStarting(totalArtifacts int)
+	BuildFinished()
+	ArtifactBuildStarting(artifactName string)
+	ArtifactBuildCached(artifactName string)
+	ArtifactBuildCompleted(artifactName string, number, total int)
+	ArtifactBuildFailed(artifactName string, errorMsg string)
+}
+
+type logRequest struct {
+	RecipeID string `json:"recipeID"`
+}
+
+type artifactMapping struct {
+	Name    *string
+	Version *string
+}
+
+type message struct {
+	Type         string      `json:"type"`
+	CacheHit     bool        `json:"cache_hit"`
+	ErrorMessage string      `json:"error_message"`
+	ArtifactID   strfmt.UUID `json:"artifact_id"`
+}
+
+func (r *Request) Wait() error {
+	url := api.GetServiceURL(api.BuildLogStreamer)
+	header := make(http.Header)
+	header.Add("Origin", "https://"+url.Host)
+
+	logging.Debug("Creating websocket for %s (origin: %s)", url.String(), header.Get("Origin"))
+	conn, _, err := websocket.DefaultDialer.Dial(url.String(), header)
+	if err != nil {
+		return errs.Wrap(err, "Could not create websocket dialer")
+	}
+	defer conn.Close()
+
+	readErr := make(chan error)
+	go r.responseReader(conn, readErr)
+
+	logging.Debug("sending websocket request")
+	request := logRequest{RecipeID: r.recipeID.String()}
+	if err := conn.WriteJSON(request); err != nil {
+		return errs.Wrap(err, "Could not write websocket request")
+	}
+
+	select {
+	case err := <-readErr:
+		return err
+	}
+}
+
+func (r *Request) responseReader(conn *websocket.Conn, readErr chan error) {
+	artifactMap, err := artifactMap(r.recipeID)
+	if err != nil {
+		readErr <- errs.Wrap(err, "Could not generate artifact map")
+	}
+
+	total := len(artifactMap)
```

Comment:
I noticed the AC: 
> Exclude artifacts with namespace builder and builder-lib from the progress indicator

That will need to be implemented still.

Reply:
That's here: https://github.com/ActiveState/cli/pull/1008/files#diff-4bf0b2d5fe8f37d210ff6cd68690150eR132

Though I see now that I missed `builder-lib`. Will add that.